File: android/ui-framework-sample/ui-framework/src/main/java/kh/mcnc/lib/components/BizLineChart.kt
package kh.mcnc.lib.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun BizLineChart(
    data: List<Float>,
    modifier: Modifier = Modifier,
    labels: List<String>? = null,
    height: Float = 240f,
    color: Color = Color(0xFF3880FF),
    strokeWidth: Float = 2f,
    showPoints: Boolean = true,
    showGrid: Boolean = true
) {
    Column(modifier = modifier.fillMaxWidth()) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(height.dp)
        ) {
            Canvas(modifier = Modifier.fillMaxSize().padding(vertical = 16.dp)) {
                if (data.isEmpty()) return@Canvas

                val max = data.maxOrNull()?.coerceAtLeast(0.001f) ?: 1f
                val min = data.minOrNull()?.coerceAtMost(0f) ?: 0f
                val range = max - min
                
                // Draw Grid
                if (showGrid) {
                    for (i in 0..4) {
                        val y = size.height * (i / 4f)
                        drawLine(
                            color = Color(0xFFF0F0F0),
                            start = Offset(0f, y),
                            end = Offset(size.width, y),
                            strokeWidth = 1.dp.toPx()
                        )
                    }
                }

                val path = Path()
                val points = mutableListOf<Offset>()

                data.forEachIndexed { index, value ->
                    val x = if (data.size > 1) {
                        size.width * (index.toFloat() / (data.size - 1))
                    } else {
                        size.width / 2
                    }
                    val y = size.height * (1f - ((value - min) / range))

                    points.add(Offset(x, y))

                    if (index == 0) {
                        path.moveTo(x, y)
                    } else {
                        path.lineTo(x, y)
                    }
                }

                // Draw line
                drawPath(
                    path = path,
                    color = color,
                    style = Stroke(width = strokeWidth.dp.toPx())
                )
            }

            // Draw Points
            if (showPoints && data.isNotEmpty()) {
                val max = data.maxOrNull()?.coerceAtLeast(0.001f) ?: 1f
                val min = data.minOrNull()?.coerceAtMost(0f) ?: 0f
                val range = max - min

                BoxWithConstraints(modifier = Modifier.fillMaxSize().padding(vertical = 16.dp)) {
                    val widthPx = constraints.maxWidth
                    val heightPx = constraints.maxHeight

                    data.forEachIndexed { index, value ->
                        val xRatio = if (data.size > 1) index.toFloat() / (data.size - 1) else 0.5f
                        val yRatio = 1f - ((value - min) / range)

                        Box(
                            modifier = Modifier
                                .offset(
                                    x = (maxWidth * xRatio) - 4.dp,
                                    y = (maxHeight * yRatio) - 4.dp
                                )
                                .size(8.dp)
                                .background(color, CircleShape)
                                .border(2.dp, Color.White, CircleShape)
                        )
                    }
                }
            }
        }

        if (!labels.isNullOrEmpty()) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 8.dp),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                labels.forEach { label ->
                    Text(
                        text = label,
                        fontSize = 12.sp,
                        color = Color(0xFF999999),
                        textAlign = TextAlign.Center,
                        modifier = Modifier.weight(1f)
                    )
                }
            }
        }
    }
}

File: ios/ui-framework-sample/UIFramework/Sources/UIFramework/BizLineChart.swift
import SwiftUI

public struct BizLineChart: View {
    public var data: [Double]
    public var labels: [String]?
    public var color: Color
    public var strokeWidth: CGFloat
    public var showPoints: Bool
    public var showGrid: Bool
    
    public init(
        data: [Double],
        labels: [String]? = nil,
        color: Color = Color(hex: "#3880FF"),
        strokeWidth: CGFloat = 2,
        showPoints: Bool = true,
        showGrid: Bool = true
    ) {
        self.data = data
        self.labels = labels
        self.color = color
        self.strokeWidth = strokeWidth
        self.showPoints = showPoints
        self.showGrid = showGrid
    }
    
    public var body: some View {
        VStack(spacing: 8) {
            GeometryReader { geometry in
                let maxVal = Swift.max(data.max() ?? 1, 0.001)
                let minVal = Swift.min(data.min() ?? 0, 0)
                let range = maxVal - minVal
                
                ZStack {
                    // Grid
                    if showGrid {
                        ForEach(0..<5) { i in
                            let y = geometry.size.height * CGFloat(i) / 4.0
                            Path { path in
                                path.move(to: CGPoint(x: 0, y: y))
                                path.addLine(to: CGPoint(x: geometry.size.width, y: y))
                            }
                            .stroke(Color(hex: "#F0F0F0"), lineWidth: 1)
                        }
                    }
                    
                    if !data.isEmpty {
                        // Line
                        Path { path in
                            for (index, value) in data.enumerated() {
                                let xRatio = data.count > 1 ? CGFloat(index) / CGFloat(data.count - 1) : 0.5
                                let yRatio = CGFloat(1.0 - ((value - minVal) / range))
                                
                                let pt = CGPoint(
                                    x: geometry.size.width * xRatio,
                                    y: geometry.size.height * yRatio
                                )
                                
                                if index == 0 {
                                    path.move(to: pt)
                                } else {
                                    path.addLine(to: pt)
                                }
                            }
                        }
                        .stroke(color, style: StrokeStyle(lineWidth: strokeWidth, lineCap: .round, lineJoin: .round))
                        
                        // Points
                        if showPoints {
                            ForEach(0..<data.count, id: \.self) { index in
                                let value = data[index]
                                let xRatio = data.count > 1 ? CGFloat(index) / CGFloat(data.count - 1) : 0.5
                                let yRatio = CGFloat(1.0 - ((value - minVal) / range))
                                
                                let pt = CGPoint(
                                    x: geometry.size.width * xRatio,
                                    y: geometry.size.height * yRatio
                                )
                                
                                Circle()
                                    .fill(color)
                                    .frame(width: 8, height: 8)
                                    .overlay(Circle().stroke(Color.white, lineWidth: 2))
                                    .position(pt)
                            }
                        }
                    }
                }
            }
            .padding(.vertical, 16)
            
            // X-Axis Labels
            if let labels = labels, !labels.isEmpty {
                HStack {
                    ForEach(0..<labels.count, id: \.self) { index in
                        Text(labels[index])
                            .font(.system(size: 12))
                            .foregroundColor(Color(hex: "#999999"))
                            .frame(maxWidth: .infinity, alignment: .center)
                    }
                }
            }
        }
    }
}
